network/requestTracker.go

```diff
@@ -0,0 +1,399 @@
+	newEntry = &TrackerRequest{
+		created:    createTime,
+		remoteHost: remoteHost,
+		remotePort: remotePort,
+		remoteAddr: remoteAddr,
+	}
+	// prune list first.
+	pruneIdx := ard.findTimestampIndex(rateLimitingWindowStartTime)
+	if pruneIdx > 0 {
+		ard.requests = ard.requests[pruneIdx:]
+	}
+	// find the new item index.
+	itemIdx := ard.findTimestampIndex(newEntry.created)
+	if itemIdx >= len(ard.requests) {
+		// it's going to be added as the last item on the list.
+		ard.requests = append(ard.requests, newEntry)
+		return
+	}
+	if itemIdx == 0 {
+		// it's going to be added as the first item on the list.
+		ard.requests = append([]*TrackerRequest{newEntry}, ard.requests...)
+		return
+	}
+	// it's going to be added somewhere in the middle.
+	ard.requests = append(ard.requests[:itemIdx], append([]*TrackerRequest{newEntry}, ard.requests[itemIdx:]...)...)
+	return
+}
+
+func (ard *hostIncomingRequests) remove(trackedRequest *TrackerRequest) {
+	for i := range ard.requests {
+		if ard.requests[i] == trackedRequest {
+			// remove entry.
+			ard.requests = append(ard.requests[0:i], ard.requests[i+1:]...)
+			return
+		}
+	}
+}
+
+func (ard *hostIncomingRequests) countConnections(rateLimitingWindowStartTime time.Time, tcpRequests bool) (count uint) {
+	i := ard.findTimestampIndex(rateLimitingWindowStartTime)
+	if tcpRequests {
+		for ; i < len(ard.requests); i++ {
+			if ard.requests[i].request == nil {
+				count++
+			}
+		}
+	} else {
+		for ; i < len(ard.requests); i++ {
+			if ard.requests[i].request != nil {
+				count++
+			}
+		}
+	}
+	return
+}
+
+// RequestTracker tracks the incoming request connections
+type RequestTracker struct {
+	downstreamHandler http.Handler
+	log               logging.Logger
+	config            config.Local
+	// once we detect that we have a misconfigured UseForwardedForAddress, we set this and write an warning message.
+	misconfiguredUseForwardedForAddress bool
+
+	listener              net.Listener
+	hostRequests          map[string]*hostIncomingRequests // maps a request host to a request data (i.e. "1.2.3.4" -> *hostIncomingRequests )
+	hostRequestsMu        deadlock.Mutex
+	acceptedConnections   map[string]*TrackerRequest // maps a remote address to a tracked request data (i.e. "1.2.3.4:1560" -> *TrackerRequest ); used to associate connection between the Accept and the ServeHTTP
+	acceptedConnectionsMu deadlock.Mutex
+}
+
+func makeRequestsTracker(downstreamHandler http.Handler, log logging.Logger, config config.Local) *RequestTracker {
+	return &RequestTracker{
+		downstreamHandler:   downstreamHandler,
+		log:                 log,
+		config:              config,
+		hostRequests:        make(map[string]*hostIncomingRequests, 0),
+		acceptedConnections: make(map[string]*TrackerRequest, 0),
+	}
+}
+
+// Accept waits for and returns the next connection to the listener.
+func (rt *RequestTracker) Accept() (conn net.Conn, err error) {
+	for {
```

Comment:
I'm not a big fan of using the goto as a language construct; naturally, the for loop is implemented with a goto, but that shouldn't be a factor.
do you think that it would be cleaner if I would move the inner logic into a separate function ( i.e. "accept" ) and leave the for just for the rejected incoming requests ?

Reply:
no need to separate the function, all of this stuff is one chunk that makes sense. probably just leave this as is.